### PR TITLE
R2 fetch helper and commands

### PR DESCRIPTION
**Pull request by JacobMGEvans**
Commands added for uploading and downloading objects from r2.

### CODE REVIEW

**Comment by JacobMGEvans** [comment]
**Showing upload with `object put`**
Screen Shot 2022-07-19 at 1 52 21 PM

**Result of `object put` in dashboard R2 bucket**
Screen Shot 2022-07-19 at 1 53 03 PM

**Showing download with `object get`**
Screen Shot 2022-07-19 at 1 54 07 PM

**Local Worker object downloaded after using `objects get`**
Screen Shot 2022-07-19 at 1 54 30 PM

**Comment by threepointone** [comment]
Wooo, exciting!

**Comment by JacobMGEvans** [comment]
I will be making a separate PR for "DELETE"

File: packages/wrangler/src/__tests__/helpers/mock-cfetch.ts [hunk 1]
```diff
@@ -206,6 +209,48 @@ export function setMockFetchKVGetValue(
...
+/**
+ * Mock setter for usage within test blocks, companion helper to `mockFetchR2Objects`
+ */
+export function setMockFetchR2Objects(
+	{
+		accountId = "r2-object-test-account",
+		bucketName = "r2-test-bucket",
+		objectName = "r2-test-object",
+	},
+	value?: string
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:245]: How about you move this parameter inside the first parameter e.g.:

```ts
export function setMockFetchR2Objects(
	{
		accountId = "r2-object-test-account",
		bucketName = "r2-test-bucket",
		objectName = "r2-test-object",
    mockResponse?: string,
	},
) { ... }
```

File: packages/wrangler/src/__tests__/helpers/mock-cfetch.ts [hunk 2]
```diff
@@ -206,6 +209,48 @@ export function setMockFetchKVGetValue(
...
+) {
+	r2GetMocks.set(
+		`/accounts/${accountId}/r2/buckets/${bucketName}/objects/${objectName}`,
+		value
+	);
+}
+
 export function unsetMockFetchKVGetValues() {
 	kvGetMocks.clear();
+	r2GetMocks.clear();
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:255]: This either needs to go in its own function, or this function needs renaming. Currently it only refers to KV values.

File: packages/wrangler/src/__tests__/helpers/mock-cfetch.ts [hunk 3]
```diff
@@ -206,6 +209,48 @@ export function setMockFetchKVGetValue(
...
+export async function mockFetchR2Objects(
+	resource: string,
+	bodyInit: {
+		body: BodyInit | Readable;
+		headers: HeadersInit | undefined;
+		method: "PUT" | "GET";
+	}
+): Promise<Response> {
+	if (bodyInit.body instanceof Readable) {
+		bodyInit.body.destroy();
```
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:224]: Definitely!

File: packages/wrangler/src/__tests__/helpers/mock-cfetch.ts (hunk 1 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:245]: Sure, I don't have strong opinion on the param structure.

File: packages/wrangler/src/__tests__/helpers/mock-cfetch.ts (hunk 2 above)
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/helpers/mock-cfetch.ts:255]: Im thinking renaming it to `unsetSpecialMockFns`

**Review by petebacondarwin — APPROVED** [review]

**Pull request merged**